[Kavan Tan] iP

kavantan opened the pull request:
# Duke
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/kavantan/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

kavantan committed:
Added .class files to .gitignore

kavantan committed:
Implemented Level-1: Greet, Echo, Exit

kavantan committed:
Moved HORIZONTAL_LINE constant into main

kavantan committed:
Stylistic changes (indentations)

kavantan committed:
Imported ArrayList and initialised storedText array

kavantan committed:
Added 'list' conditional logic and implementation

kavantan committed:
Stylistic changes to match requirements (list numbering, 'added:')

kavantan committed:
Created Task.java skeleton class

kavantan committed:
Added markAsDone method in Task class

kavantan committed:
Added markAsUndone method in Task class

kavantan committed:
Added toString method to Task class

kavantan committed:
Changed Array to store instances of class Task instead of String

kavantan committed:
Implemented 'mark' logic

kavantan committed:
Implemented 'unmark' logic

kavantan committed:
Added Todos class which inherits Task class

kavantan committed:
Added Deadlines class which inherits Task class

kavantan committed:
Added Events class which inherits Task class

kavantan committed:
Added 'todo' functionality into main logic

kavantan committed:
Stylistic change to match requirements (added space to Deadlines' toString method)

kavantan committed:
Stylistic change to match requirements (added whitespaces to Deadlines' and Events' toString methods respectively)

kavantan committed:
Added 'deadline' functionality into main logic

kavantan committed:
Added 'event' functionality into main logic

kavantan committed:
Created DukeException class which inherits Exception

kavantan committed:
Improved String concatenation using StringBuilder

kavantan committed:
Slightly refactored Duke.java and added 'unknown input' DukeException

kavantan committed:
Rectified bug in 'todo' conditional

kavantan committed:
Added 'empty todo input' DukeException

kavantan committed:
Added 'delete' functionality into tracker logic

kavantan committed:
Refactored Duke.java to a greater extent

kavantan committed:
Refactored DukeException.java to include the 'oops' part

kavantan committed:
Implemented testing using the i/o redirection technique

kavantan committed:
Change taskNumber type declaration to 'int' from 'Integer'

kavantan committed:
Implement StoredTasks.java with skeleton 'save' method

kavantan committed:
Implement skeleton 'load' method in StoredTasks.java

kavantan committed:
Add data subfolder to .gitignore (Prevent git tracking of stored tasks)

kavantan committed:
Change Task class to be abstract and added abstract storedTaskString method

kavantan committed:
Implement storedTaskString method in Todos.java, Deadlines.java and Deadlines.java

kavantan committed:
Implement (fully) load and save methods in StoredTasks.java

kavantan committed:
Implement load and save capabilities into Duke.java

kavantan committed:
Remove redundant imports from Duke.java

kavantan committed:
Add LocalDate encapsulation into Deadline class

kavantan committed:
Implement parsing of string into LocalDate object in Duke.java

kavantan committed:
Refactor 'date and time parsing logic' into its own class

kavantan committed:
Merge branch 'branch-Level-8'

kavantan committed:
Resolve incompatibility issues with branch-Level-7 and branch-Level-8

kavantan committed:
Add constructor for StoredTasks class

kavantan committed:
Implement Ui class which deals with user interactions

kavantan committed:
Implement Parser class which handles user commands

kavantan committed:
Refactor Duke to encapsulate Ui, Parser and StoredTasks classes

kavantan committed:
Implement TaskList class which contains the task list

kavantan committed:
Refactor load functionality from StoredTasks to TaskList

kavantan committed:
Refactor Duke and Parser to encapsulate TaskList

kavantan committed:
Put all classes into duke package

kavantan committed:
Fix incorrect indentations across all project files

kavantan committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

kavantan committed:
Organise classes into more packages duke.task and duke.util

kavantan committed:
Implement DateAndTimeParserTest using Junit

kavantan committed:
Correct build.gradile file values w.r.t Shadow plugin

kavantan committed:
Add JavaDoc comments to all non-private classes/methods

kavantan committed:
Tweak code to comply with given coding standard

kavantan committed:
Implement containsKeyword method in Task class

kavantan committed:
Add 'find' functionality into Parser class

kavantan committed:
Rectify TaskList class code to comply with given coding standard

kavantan committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java

kavantan committed:
Modify Array usage in containsKeyword method

kavantan committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/Task.java

kavantan committed:
Add JavaDoc comment to containsKeyword method

kavantan committed:
Configure project to use CheckStyle

qiaoen17 commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,102 @@
+        } else if (command.equals("list")) {
+            System.out.println(HORIZONTAL_LINE);
+            for (int i = 0; i < storedTasks.size(); i++) {
+                System.out.println("  " + String.valueOf(i + 1) + ". " + storedTasks.get(i));
```
Should this be extracted out to the Ui class? I noticed this in the other lines too.

qiaoen17 commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,102 @@
+            storedTasks.add(todo);
+            System.out.println(
+                    HORIZONTAL_LINE + "\n  Got it. I've added this task:\n  " + todo + "\n  Now you have "
+                            + String.valueOf(storedTasks.size()) + " tasks in the list.\n" + HORIZONTAL_LINE);
```
Should this be extracted to the ToDo class or TaskList class instead? I noticed this for the deadlines and event class too.

qiaoen17 commented on src/main/java/StoredTasks.java:
```diff
@@ -21,7 +24,30 @@ public static ArrayList<Task> load() {
+                    }
+                    storedTasks.add(event);
+                    break;
+                }
```
I like this use of switch cases :)

qiaoen17 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,85 @@
+        return saveString.toString();
+    }
+
+    private ArrayList<Task> load() throws DukeException {
```
Should this function be in your StoredTasks class instead?

qiaoen17 reviewed (commented):
I think it was a good effort :)

limweijun commented:
Your code looks pretty good

kavantan committed:
Set up first JavaFX program

kavantan committed:
Implement 'echoing the user' functionality

kavantan committed:
Add DaUser and DaDuke images

kavantan committed:
Implement custom dialogBox for Duke

kavantan committed:
Add DialogBox and MainWindow fxml files to rebuild Scene

kavantan committed:
Refactor Duke to incorporate fxml

kavantan committed:
Refactor Parser into individual command subclasses

kavantan committed:
Change String representation in Ui slightly

kavantan committed:
Refactor runDuke() into new getResponse() method

kavantan committed:
Remove text-based UI

kavantan committed:
Add JavaDoc comments to new refactored classes and methods

kavantan committed:
Edit build.gradle to contain correct values w.r.t. Shadow plugin

kavantan committed:
Add assertions

Codebase has no assertions.

The lack of assertions means that there is not enough error handling.

Adding assert statements allows us to document assumptions that should hold at
various points in the code.

kavantan committed:
Remove Ui.java and its instances

Duke class creates a Ui object instance.

This is not necessary because Duke has moved on from the text-based UI
to the GUI.

Removing Ui class and its instances eliminates potential confusion as to
the Ui class' purpose.

kavantan committed:
Merge pull request #2 from kavantan/branch-A-Assertions

Add assertions

kavantan committed:
Merge pull request #3 from kavantan/branch-A-CodeQuality

Remove Ui.java and its instances

kavantan committed:
Implement feature to update task description

kavantan committed:
Merge pull request #4 from kavantan/branch-C-Update

Implement feature to update task description

kavantan committed:
Add Product Screenshot file (Ui.png) into docs

kavantan committed:
Resolve 'master' branch merge conflicts

kavantan committed:
Update User Guide for Duke
